scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py

```diff
@@ -139,17 +139,46 @@ def _merge_workspace_with_limits(focused_ws, q_lims):
     min_x = np.inf
     max_x = -np.inf
     num_x = -np.inf
-    for ws in focused_ws:
-        x_data = ws.dataX(0)
+    ws_max_range = 0
+    largest_range_spectrum = 0
+    for i in range(focused_ws.size()):
+        x_data = focused_ws[i].dataX(0)
         min_x = min(np.min(x_data), min_x)
         max_x = max(np.max(x_data), max_x)
         num_x = max(x_data.size, num_x)
-    width_x = (max_x-min_x)/num_x
-    binning = [min_x, width_x, max_x]
-    focused_ws = mantid.Rebin(InputWorkspace=focused_ws, Params=binning)
-    focused_ws_conjoined = mantid.ConjoinSpectra(InputWorkspaces=focused_ws)
+        ws_range = np.max(x_data)-np.min(x_data)
+        if ws_range > ws_max_range:
+            largest_range_spectrum = i + 1
+            ws_max_range = ws_range
+    focused_ws = mantid.Rebin(InputWorkspace=focused_ws, Params=[min_x, (max_x-min_x)/num_x, max_x])
+    while focused_ws.size() > 1:
+        mantid.ConjoinWorkspaces(InputWorkspace1=focused_ws[0],
+                                 InputWorkspace2=focused_ws[1])
+    focused_ws_conjoined = focused_ws[0]
+    mantid.MatchSpectra(InputWorkspace=focused_ws_conjoined, OutputWorkspace=focused_ws_conjoined,
+                        ReferenceSpectrum=largest_range_spectrum)
+
+    q_min, q_max = _load_qlims(q_lims)
+    bin_width = np.inf
+    for i in range(q_min.size):
+        pdf_x_array = focused_ws_conjoined.readX(i)
+        tmp1 = np.where(pdf_x_array >= q_min[i])
+        tmp2 = np.amin(tmp1)
+        q_min[i] = pdf_x_array[tmp2]
+        q_max[i] = pdf_x_array[np.amax(np.where(pdf_x_array <= q_max[i]))]
+        bin_width = min(pdf_x_array[1] - pdf_x_array[0], bin_width)
```

Comment:
Ditto on hoisting out and unit testing this loop to get q_min and q_max

And handling np.inf case if that propagates through below